Add LSP tracing

AleksanderBG opened the pull request:
LSP communication can now be traced remotely.

See https://github.com/lampepfl/dotty-remote-tracer for server implementation.

Duhemm commented on vscode-dotty/src/tracer.ts:
```diff
@@ -0,0 +1,281 @@
+                'Do you want to help EPFL develop this plugin by uploading your usage data? ' +
+                'PLEASE BE AWARE that this will upload all of your keystrokes and all of your code, ' +
+                'among other things.',
+                'yes', 'no'
```
I think that just saying "among other things" is scary. We should either exhaustively list what's being uploaded here or include a link to a page that exhaustively lists what's being uploaded.

Also, I would replace "all of your keystrokes" by "every action performed in Dotty IDE". Keystrokes outside of the IDE are not recorded.

AleksanderBG commented:
@Duhemm fixed most of the things you mentioned. Tracing popup now directly mentions that we simply upload all LSP communication, which contains all the files and all the keystrokes. Added a statusbar item which tells the user tracing is on and allows them to adjust their consent.

There are still some outlying issues (Dotty codebase cannot be really traced, b/c the files it contains are too large; websocket connections get dropped if internet connection disappers for a second and cannot be recreated because the server does not support it; we should probably clean the LSP output channel from time to time to avoid too much trash, but do so in a way consistent with other plugins), but this was meant to work for students and it should be sufficient for them. @smarter?

smarter commented:
> Dotty codebase cannot be really traced, b/c the files it contains are too large

What's the exact issue here? Can we work around it by sending compressed files ?

AleksanderBG commented:
@smarter I think I found a good solution. The problem starts with the fact that LSP sometimes sends entire files in one message, at least when the file is opened. This causes issues, since there's a limit on the size of a message sent through WebSocket (probably both in Play Framework and Nginx), and it may not be possible (or desireable) to completely remove it. Allowing arbitrary size messages may open the server to easy DoS, and setting a limit on the size of message will ultimately just delay the problem.

The solution would be to send messages in chunks if they are too large, and save each chunk to disk immediately after it is received. This avoid the problem of too large messages, but creates a possibility where the last line of the log on the disk is malformed (because not all chunks were sent/received). I think we can work around that limitation more easily, since it's enough to just check if the file ends in a newline.

AleksanderBG commented:
Rebased, updated `npm` to a version which actually notices and uses the `package-lock.json`, added a setting for maximum message size. Could not get the connection to fail while opening files in Dotty repo.

smarter commented on vscode-dotty/src/tracer.ts:
```diff
@@ -0,0 +1,370 @@
+        this.sessionId = new Date().toISOString()
+    }
+
+    run(): { lspOutputChannel?: vscode.OutputChannel } {
```
Why not `run(): vscode.OutputChannel | undefined` ?

smarter commented on vscode-dotty/src/tracer.ts:
```diff
@@ -0,0 +1,370 @@
+
+    private initializeAsyncWorkspaceDump() {
+        if (this.remoteWorkspaceDumpUrl === undefined) return
+        // convince TS that this is a string
```
You can convince TS of this in a more direct way:
```typescript
const url = this.remoteWorkspaceDumpUrl
if (url) {
  // ...
}
```

smarter commented on vscode-dotty/src/tracer.ts:
```diff
@@ -0,0 +1,370 @@
+                        type: event.type
+                    }))
+                )
+                vscode.window.showWarningMessage('An error occured in Dotty LSP remote tracing connection.')
```
I would rather not show any messages to users if something goes wrong, it's distracting and not very useful for them. What we could do is change the status bar message to have On/Off/Failed

smarter commented on vscode-dotty/src/tracer.ts:
```diff
@@ -0,0 +1,370 @@
+    value: T
+    valueToObject: (t: T) => {}
+
+    constructor(value: T, valueToObject: (t: T) => {}) {
```
You can directly create an immutable class field and assign by writing `constructor(readonly value: T, ...)`

smarter commented on vscode-dotty/src/tracer.ts:
```diff
@@ -0,0 +1,370 @@
+    return new SafeJsonifier(e, (e) => e.toString())
+}
+
+class SafeJsonifier<T> {
```
Some documentation would be nice :)

smarter commented on vscode-dotty/src/tracer.ts:
```diff
@@ -0,0 +1,370 @@
+
+        this.tracingConsent = new TracingConsentCache(ctx.extensionContext.workspaceState)
+
+        this.remoteWorkspaceDumpUrl = this.ctx.extensionConfig.get<string>('remoteWorkspaceDumpUrl')
```
I think these should be in "tracing." like maximumMessageSize

smarter commented on vscode-dotty/src/tracer.ts:
```diff
@@ -0,0 +1,370 @@
+    }
+
+    private createLspOutputChannel(): vscode.OutputChannel | undefined {
+        if (!this.remoteTracingUrl) return undefined
```
Unless I'm missing something, it looks like this means that we don't get any outputchannel, even a local one, when remoteTracingUrl is undefined. Suggestion: make an LspOutputChannel that extends OutputChannel and take as argument a (local) OutputChannel, then every method of LspOutputChannel just forwards to the underlying OutputChannel and also does network stuff if tracing is enabled.
